### ISSUE TITLE
Simple Hello-World BOF crashes beacon

### ISSUE THREAD

**Issue by 1mansh0w** [BishopFox/sliver#794]
I'm having issues creating a simple BOF that does not crash the beacon.

Here is the source code:
https://github.com/1mansh0w/sliver-bof-hello-world

Here is how I load it:
```
1. extensions rm say-hello
2. extensions load "[path]/hello_world"
3. extensions install "[path]/hello_world"  
```

When I now execute the BOF **only once it always runs fine**, but executing it a couple of more times will **always crash the beacon**. The client will just show this message and the session is lost:

`"Executing say-hello ..."`

**Am I doing something wrong or is there a bug in sliver?** 

Thanks!

___
I'm running the latest version of Sliver (v1.5.22) on both client and server. I'm cross-compiling the BOF on Ubuntu.

**Comment by rkervella** [comment]
Are you running a beacon or a session?

**Comment by rkervella** [comment]
Yup I've seen that happens with other BOFs, as stated in #679. I'll have a look. I looked at your BOF, it seems fine so it's definitely a bug on our side.

**Comment by rkervella** [comment]
The annoying part is I can't seem to be able to reproduce:

```shell
[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World
[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello
[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World

[server] sliver (TIRED_BUSTLE) > say-hello

[*] Successfully executed say-hello (coff-loader)
[*] Got output:
Hello World
```

**Comment by 1mansh0w** [comment]
Interesting...

Is there any information/log when executing BOFs? Like a stack trace or any information that helps debugging once a session/beacon crashes?

**Comment by rkervella** [comment]
Yes, if you generate your implant with `--debug`, and run it from a terminal, you should be able to see the crash.

**Comment by r00t0v3rr1d3** [comment]
sa-uptime and sa-resources when in beacon mode are consistent in eventually crashing for me

**Comment by rkervella** [comment]
Only in beacon mode though?

**Comment by r00t0v3rr1d3** [comment]
For me, MOST of the instability / crashing seems to go away in session mode. Occasionally I can get those same BOFs to hang forever in session mode if I up arrow + enter really fast.

**Comment by moloch--** [comment]
Hurm, i wonder if its some race condition in the invocation?

**Comment by rkervella** [comment]
Yeah definitely looks like it

**Comment by 1mansh0w** [comment]
What's really weird is now that I have created an implant with `--debug` it doesn't crash anymore. I cannot get my _say-hello_ BOF to crash and neither _sa-uptime_:

![Screen Shot 2022-08-22 at 18 56 33](https://user-images.githubusercontent.com/91868469/185976965-c7c667cf-84ae-4b62-b45b-ea3182eb07b5.png)

So I thought there was an issue with the old implant, but when I create a new implant (v1.5.22) **without** `--debug` I can get it crashing again:

![Screen Shot 2022-08-22 at 18 25 33](https://user-images.githubusercontent.com/91868469/185971379-118fb641-03a9-4124-812b-3edeaf68a99d.png)

![Screen Shot 2022-08-22 at 18 51 23](https://user-images.githubusercontent.com/91868469/185976306-0398cce4-df88-4f28-a455-04d25b98ffba.png)

I didn't look into the code yet to see if the `--debug ` flag could really be the problem. @r00t0v3rr1d3 Do you see any difference when you build an implant with the debug flag?

**Comment by rkervella** [comment]
Hm that's good to know. I'll run more tests with obfuscation enabled, see if crashes. If so, it could also be an issue with garble.

**Comment by r00t0v3rr1d3** [comment]
I haven’t tried recently to build one using the —debug flag. Not with Golang, but with C - I’ve definitely seen things where the debug version magically saved crashes from happening that the exact same code in “release” form would cause it to fall on its face.

**Comment by rkervella** [comment]
Alright I can reproduce with obfuscation enabled.

**Comment by rkervella** [comment]
Alright it was indeed a race condition, PR #797 should fix this one and #679 .
